src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1174 @@
+
+        var frames_metadata = try FramesMetadata.init(init_allocator, num_frames);
+        errdefer frames_metadata.deinit(init_allocator);
+
+        var free_list = try AtomicStack(FrameIndex).init(init_allocator, num_frames);
+        errdefer free_list.deinit(init_allocator);
+        for (0..num_frames) |i| free_list.appendAssumeCapacity(@intCast(i));
+
+        var io_uring = if (builtin.os.tag == .linux) blk: {
+            // NOTE: this is pretty much a guess, maybe worth tweaking?
+            // think this is a bit on the high end, libxev uses 256
+            const io_uring_entries = 4096;
+
+            break :blk try std.os.linux.IoUring.init(
+                io_uring_entries,
+                0,
+            );
+        } else {};
+        errdefer if (builtin.os.tag == .linux) io_uring.deinit();
+
+        var frame_map: FrameMap = .{};
+        try frame_map.ensureTotalCapacity(init_allocator, num_frames);
+        errdefer frame_map.deinit(init_allocator);
+
+        return .{
+            .frames = frames,
+            .frames_metadata = frames_metadata,
+            .free_list = free_list,
+            .frame_map = frame_map,
+            .eviction_lfu = try HierarchicalFIFO.init(init_allocator, num_frames / 10, num_frames),
+            .io_uring = io_uring,
+        };
+    }
+
+    pub fn deinit(self: *BufferPool, init_allocator: std.mem.Allocator) void {
+        init_allocator.free(self.frames);
+        self.frames_metadata.deinit(init_allocator);
+        if (builtin.os.tag == .linux) {
+            self.io_uring.deinit();
+        }
+        self.free_list.deinit(init_allocator);
+        self.eviction_lfu.deinit(init_allocator);
+        self.frame_map.deinit(init_allocator);
+    }
+
+    pub fn computeNumberofFrameIndices(
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) u32 {
+        if (file_offset_start > file_offset_end) unreachable;
+        if (file_offset_start == file_offset_end) return 0;
+
+        const starting_frame = file_offset_start / FRAME_SIZE;
+        const ending_frame = (file_offset_end - 1) / FRAME_SIZE;
+
+        return ending_frame - starting_frame + 1;
+    }
+
+    /// allocates the required amount of indices, sets them all to
+    /// INVALID_FRAME, overwriting with a valid frame where one is found.
+    /// INVALID_FRAME indicates that there is no frame in the BufferPool for the
+    /// given file_id and range.
+    fn computeFrameIndices(
+        self: *BufferPool,
+        file_id: FileId,
+        allocator: std.mem.Allocator,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) error{ InvalidArgument, OutOfMemory }![]FrameIndex {
+        const n_indices = computeNumberofFrameIndices(file_offset_start, file_offset_end);
+
+        if (n_indices > self.frames.len) return error.InvalidArgument;
```

Comment:
```suggestion
        if (n_indices > self.frames.len) return error.OffsetsOutOfBounds;
```

Reply:
Changed name in 25ac1b13744cdd660b4643481fc848c7e7dd4532